File: Apps/Web/src/types/market.ts
export type MarketSource = "binance" | "coingecko";
export type RSISignal = "oversold" | "neutral" | "overbought";
export type MACDTrend = "bullish" | "bearish" | "neutral";
export type Interval = "4h" | "1d" | "1w" | "1m" | "1y";

export interface MarketTick {
  symbol: string;
  price: string;
  change_24h: number;
  change_pct_24h: number;
  volume_24h: string;
  high_24h: string;
  low_24h: string;
  market_cap: string | null;
  timestamp: string;
  source: MarketSource;
}

export interface OHLCVCandle {
  open_time: string;
  open: string;
  high: string;
  low: string;
  close: string;
  volume: string;
  close_time: string;
  quote_volume: string | null;
  num_trades: number | null;
}

export interface RSIResult {
  value: number;
  period: number;
  signal: RSISignal;
}

export interface MACDResult {
  macd_line: number;
  signal_line: number;
  histogram: number;
  trend: MACDTrend;
}

export interface EMAResult {
  value: number;
  period: number;
}

export interface SMAResult {
  value: number;
  period: number;
}

export interface BollingerResult {
  upper: number;
  middle: number;
  lower: number;
  bandwidth: number;
  percent_b: number;
}

export interface IndicatorSet {
  symbol: string;
  interval: string;
  rsi: RSIResult | null;
  macd: MACDResult | null;
  ema_20: EMAResult | null;
  ema_50: EMAResult | null;
  sma_200: SMAResult | null;
  bollinger: BollingerResult | null;
  computed_at: string;
}

export interface MarketPricesResponse {
  items: MarketTick[];
  count: number;
}

export interface OHLCVResponse {
  symbol: string;
  interval: string;
  items: OHLCVCandle[];
  count: number;
}

export interface SearchResult {
  id: string;
  symbol: string;
  name: string;
  market_cap_rank: number | null;
  thumb: string | null;
}

File: Apps/Api/schemas/market.py
from __future__ import annotations
from datetime import datetime
from decimal import Decimal
from typing import Literal, List
from pydantic import BaseModel, Field


class MarketTick(BaseModel):
    symbol: str
    price: Decimal
    change_24h: float
    change_pct_24h: float
    volume_24h: Decimal
    high_24h: Decimal
    low_24h: Decimal
    market_cap: Decimal | None = None
    timestamp: datetime
    source: Literal["binance", "coingecko"]


class OHLCVCandle(BaseModel):
    open_time: datetime
    open: Decimal
    high: Decimal
    low: Decimal
    close: Decimal
    volume: Decimal
    close_time: datetime
    quote_volume: Decimal | None = None
    num_trades: int | None = None


class RSIResult(BaseModel):
    value: float
    period: int
    signal: Literal["oversold", "neutral", "overbought"]


class MACDResult(BaseModel):
    macd_line: float
    signal_line: float
    histogram: float
    trend: Literal["bullish", "bearish", "neutral"]


class EMAResult(BaseModel):
    value: float
    period: int


class SMAResult(BaseModel):
    value: float
    period: int


class BollingerResult(BaseModel):
    upper: float
    middle: float
    lower: float
    bandwidth: float
    percent_b: float


class IndicatorSet(BaseModel):
    symbol: str
    interval: str
    rsi: RSIResult | None = None
    macd: MACDResult | None = None
    ema_20: EMAResult | None = None
    ema_50: EMAResult | None = None
    sma_200: SMAResult | None = None
    bollinger: BollingerResult | None = None
    computed_at: datetime = Field(default_factory=datetime.utcnow)


class MarketPricesResponse(BaseModel):
    items: List[MarketTick]
    count: int


class OHLCVResponse(BaseModel):
    symbol: str
    interval: str
    items: List[OHLCVCandle]
    count: int


class SearchResult(BaseModel):
    id: str
    symbol: str
    name: str
    market_cap_rank: int | None = None
    thumb: str | None = None
